# Assignment: Programming Review
## Do Q1 and one other question.

**Q1.** First, think about your priorities in life. What kind of salary do you want to make after graduation? Do you mind getting more schooling? What kind of work-life balance are you looking for? Where do you want to work, geographically? You don't have to write this down here, just think about it.

1. Go to the Occupational Outlook Handbook at [https://www.bls.gov/ooh/](https://www.bls.gov/ooh/). Look up "Data Scientist." Read about the job and start collecting data about it from the job profile (e.g. salary, education required, work setting)
2. Find 7-10 other jobs that appeal to you, and collect the same data as you did for Data Scientist. Put it all in a spreadsheet.
3. Do any of your findings surprise you?
4. Rank the jobs you picked from best to worst, and briefly explain why you did so.
5. Please submit your spreadsheet with the assignment --- you can "de-identify" it and remove anything that you find personally identifying or you don't want to share, of course. We'll play with these data later.


ANSWERS

1. Dentist
2. Data Scientist
3. Surgeon
4. Lawyer
5. Public Relations Manager
6. Architect
7. Private Investigator
8. Middle School Teacher


I did my rankings considering median pay, job outlook, education level, and employment change. So for example, I thought that when considering each of these categories, even though middle school teachers median pay is higher than that of a PI, they have a lower required education level and a higher job outlook than that of a middle school teacher, which in my mind, allows a PI to be ranked higher.

Depends on student opinions.

**Q3.** This is a basic review of some statistics along with practice writing functions. Like we talked about, Python is a general purpose programming language and ships without basic data handling or statistical packages coded in. The beginning of the code chunk below generates random values for you to test your work on; since the random values are generates as NumPy arrays, you'll need to use the Numpy methods `np.sum(x)` to sum the vector $x$ and `np.sqrt(x)` to take the square roots of the values in $x$, as well as the Python function `len(x)` to get the length of $x$.

Try to reuse the functions you've already defined as you work through the following questions, rather than rewriting code you've already written.

1. Write a function that computes the **sample average** or **mean** of a vector $x$,
$$
\bar{x} = \dfrac{x_1 + x_2 + ... + x_N}{N} = \dfrac{\sum_{i=1}^N x_i}{N}.
$$
Write a function in the code chunk below to compute this quantity, and then use it to compute the mean of $x$.
2. Write a function that computes the **sample standard deviation** of a vector $x$,
$$
s_x = \sqrt{\dfrac{(x_1 - \bar{x})^2 + ... (x_N - \bar{x})^2 }{N-1}} = \sqrt{\dfrac{ \sum_{i=1}^N (x_i - \bar{x})^2 }{N-1}}.
$$
The intuition of this quantity is that it computes roughly the average distance from each point $x_i$ to the sample mean $\bar{x}$. If it is small, it means all the points are clustered tightly around the mean, and if it is large, it means the points are typically far away from the average. Write a function in the code chunk below to compute this quantity, and then use it to compute the sample standard deviation of $x$.
3. Write a function that calls the previous two to **standardize** the values of the vector as a **$z$-score**:
$$
z = \dfrac{x-\bar{x}}{s}.
$$
The intuition of this quantity is that it is recentering all the values of $x$ so the average is zero and then scaling them by the standard deviation. If the data are normally distributed and $N$ is large, the $z$ score will approximately follow a standard normal distribution. Write a function in the code chunk below to compute this quantity, and then use it to compute the z-scores for $x$.
4. The **sample covariance** of two vectors $x=(x_1,...,x_N)$ and $y=(y_1,...,y_N)$ is defined as
$$
cov(x,y) = \dfrac{(x_1 - \bar{x})(y_1-\bar{y}) + (x_2 - \bar{x})(y_2-\bar{y}) + ... + (x_N - \bar{x})(y_{N}-\bar{y})}{N-1}
$$
$$
= \dfrac{\sum_{i=1}^N (x_i - \bar{x})(y_i - \bar{y})}{N-1}.
$$
The intuition of this quantity is that it looks at the pairs $(x_i, y_i)$ and compares them to the means $(\bar{x},\bar{y})$ to determine whether $x$ and $y$ tend to co-vary in the same direction relative to their means: If the values of $x$ and $y$ are typically both above or below the mean values of $x$ and $y$, then $x$ and $y$ will have a positive covariance, but if $x$ is typically above the mean of $x$ when $y$ is typically below the mean of $y$ or vice versa, then they will have a negative covariance. Write a function in the code chunk below to compute this quantity, and then use it to compute the covariance of the generated $x$ and $y$.
6. The **sample correlation coefficient** of two vectors $x$ and $y$ is defined as
$$
r_{xy} = \dfrac{cov(x,y)}{s_x s_y}
$$
Use your functions to create a new function that compute this quantity. The intuition of this quantity is that it is like the covariance, but normalized so that its values like between -1 and 1: perfect negative linear association between the variables at -1, no association at 0, and perfect positive linear association between the variables at 1.

In [12]:
import numpy as np
import math as math
np.random.seed(100) # Set the seed for the random number generator
rho, sigma_x, sigma_y = -.4, 3, 2 # Variance-Covariance Parameters
vcv = np.array([[sigma_x**2, rho*sigma_x*sigma_y],
                [rho*sigma_x*sigma_y,sigma_y**2]]) # VCV Matrix
mu = np.array([-1,2]) # Population averages
sample = np.random.multivariate_normal(mu,vcv,200) # Multivariate normal draws
x = sample[:,0]
y = sample[:,1]

#############################################################################

#1 Function to compute the sample mean
def mean(x):
    m = np.sum(x)/len(x)
    return(m)
print('Sample Mean:', mean(x),'\n')


Sample Mean: -1.0899806430403438 



In [15]:
#PART 2

def standard_deviation(x):
    N = len(x)
    m = mean(x)
    sumsqdiff = np.sum( (x-m)**2 )
    sd = np.sqrt(sumsqdiff/(N-1))
    return(sd)
print('Sample Standard Deviation:', standard_deviation(x),'\n')

Sample Standard Deviation: 3.2847981697718236 



In [16]:
#PART 3

def zscore(x):
    z = (x-mean(x))/standard_deviation(x)
    return(z)
print('Z-scores:', zscore(x)[1:10],'\n')

Z-scores: [-1.05060967 -0.74576403 -0.38160532  0.24681908  0.52189922  0.70940228
 -0.59306576  0.70489592  0.19667971] 



In [17]:
#PART 4
def cov(x,y):
    N = len(x)
    sumsqdiff = np.sum( (x-mean(x))*(y-mean(y)) )
    c = sumsqdiff / (N-1)
    return(c)
print('Covariance:', cov(x,y),'\n')

Covariance: -2.8396505928938773 



In [18]:
#PART 5
def corr(x,y):
    cv = cov(x,y)
    s_x = sd(x)
    s_y = sd(y)
    r = cv/(s_x*s_y)
    return(r)
print('Correlation:', corr(x,y),'\n')

Correlation: -0.42205093921264875 



ANSWERS

Sample Mean: -1.0899806430403438

Sample Standard Deviation: 3.2847981697718236

Z-scores: [-1.05060967 -0.74576403 -0.38160532  0.24681908  0.52189922  0.70940228
 -0.59306576  0.70489592  0.19667971]

Covariance: -2.8396505928938773

Correlation: -0.42205093921264875